src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+            }
+            return null;
+        }
+
+        public override BoundNode? VisitAssignmentOperator(BoundAssignmentOperator node)
+        {
+            base.VisitAssignmentOperator(node);
+            if (node.Left.Kind != BoundKind.DiscardExpression)
+            {
+                ValidateAssignment(node.Syntax, node.Left, node.Right, node.IsRef, _diagnostics);
+            }
+            return null;
+        }
+
+        public override BoundNode? VisitIsPatternExpression(BoundIsPatternExpression node)
+        {
+            using var _ = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            return base.VisitIsPatternExpression(node);
+        }
+
+        public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+        {
+            SetLocalScopes(node);
+
+            using var _ = new PatternInput(this, getDeclarationValEscape(node.DeclaredType, _patternInputValEscape));
+            return base.VisitDeclarationPattern(node);
+
+            static uint getDeclarationValEscape(BoundTypeExpression typeExpression, uint valEscape)
+            {
+                return typeExpression.Type.IsRefLikeType ? valEscape : Binder.CallingMethodScope;
+            }
+        }
+
+        public override BoundNode? VisitListPattern(BoundListPattern node)
+        {
+            SetLocalScopes(node);
+            return base.VisitListPattern(node);
+        }
+
+        public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
+        {
+            SetLocalScopes(node);
+            return base.VisitRecursivePattern(node);
+        }
+
+        public override BoundNode? VisitPositionalSubpattern(BoundPositionalSubpattern node)
+        {
+            using var _ = new PatternInput(this, getPositionalValEscape(node.Symbol, _patternInputValEscape));
+            return base.VisitPositionalSubpattern(node);
+
+            static uint getPositionalValEscape(Symbol? symbol, uint valEscape)
+            {
+                return symbol is null
+                    ? valEscape
+                    : symbol.GetTypeOrReturnType().IsRefLikeType() ? valEscape : Binder.CallingMethodScope;
+            }
+        }
+
+        public override BoundNode? VisitPropertySubpattern(BoundPropertySubpattern node)
+        {
+            using var _ = new PatternInput(this, getMemberValEscape(node.Member, _patternInputValEscape));
+            return base.VisitPropertySubpattern(node);
+
+            static uint getMemberValEscape(BoundPropertySubpatternMember? member, uint valEscape)
+            {
+                if (member is null) return valEscape;
+                valEscape = getMemberValEscape(member.Receiver, valEscape);
+                return member.Type.IsRefLikeType ? valEscape : Binder.CallingMethodScope;
+            }
+        }
+
+        private void SetLocalScopes(BoundObjectPattern pattern)
```

Comment:
>SetLocalScopes [](http://example.com/codeflow?start=21&length=14)

nit: consider renaming this to avoid ambiguity with other `SetLocalScopes` method. This could be `SetPatternLocalScope` or some such.